akka-distributed-data/src/main/scala/akka/cluster/ddata/DeltaPropagationSelector.scala

```diff
@@ -0,0 +1,173 @@
+
+import scala.collection.immutable
+import akka.cluster.ddata.Replicator.Internal.DeltaPropagation
+import akka.actor.Address
+import akka.cluster.ddata.Replicator.Internal.DataEnvelope
+
+/**
+ * INTERNAL API: Used by the Replicator actor.
+ * Extracted to separate trait to make it easy to test.
+ */
+private[akka] trait DeltaPropagationSelector {
+
+  private var _propagationCount = 0L
+  def propagationCount: Long = _propagationCount
+  private var deltaCounter = Map.empty[String, Long]
+  private var deltaEntries = Map.empty[String, immutable.TreeMap[Long, ReplicatedData]]
+  private var deltaSentToNode = Map.empty[String, Map[Address, Long]]
+  private var deltaNodeRoundRobinCounter = 0L
+
+  def divisor: Int
+
+  def allNodes: Vector[Address]
+
+  def createDeltaPropagation(deltas: Map[String, ReplicatedData]): DeltaPropagation
+
+  def update(key: String, delta: ReplicatedData): Unit = {
+    val c = deltaCounter.get(key) match {
+      case Some(c) ⇒ c
+      case None ⇒
+        deltaCounter = deltaCounter.updated(key, 1L)
+        1L
+    }
+    val deltaEntriesForKey = deltaEntries.getOrElse(key, immutable.TreeMap.empty[Long, ReplicatedData])
+    val updatedEntriesForKey =
+      deltaEntriesForKey.get(c) match {
+        case Some(existingDelta) ⇒
+          deltaEntriesForKey.updated(c, existingDelta.merge(delta.asInstanceOf[existingDelta.T]))
+        case None ⇒
+          deltaEntriesForKey.updated(c, delta)
+      }
+    deltaEntries = deltaEntries.updated(key, updatedEntriesForKey)
+  }
+
+  def delete(key: String): Unit = {
+    deltaEntries -= key
+    deltaCounter -= key
+    deltaSentToNode -= key
+  }
+
+  def nodesSliceSize(allNodesSize: Int): Int = {
+    // 2 - 10 nodes
+    math.min(math.max((allNodesSize / divisor) + 1, 2), math.min(allNodesSize, 10))
+  }
+
+  def collectPropagations(): Map[Address, DeltaPropagation] = {
+    _propagationCount += 1
+    val all = allNodes
+    if (all.isEmpty)
+      Map.empty
+    else {
+      // For each tick we pick a few nodes in round-robin fashion, 2 - 10 nodes for each tick.
+      // Normally the delta is propagated to all nodes within the gossip tick, so that
+      // full state gossip is not needed.
+      val sliceSize = nodesSliceSize(all.size)
+      val slice = {
+        if (all.size <= sliceSize)
+          all
+        else {
+          val i = (deltaNodeRoundRobinCounter % all.size).toInt
+          val first = all.slice(i, i + sliceSize)
+          if (first.size == sliceSize) first
+          else first ++ all.take(sliceSize - first.size)
+        }
+      }
+      deltaNodeRoundRobinCounter += sliceSize
+
+      var result = Map.empty[Address, DeltaPropagation]
+
+      slice.foreach { node ⇒
+        // collect the deltas that have not already been sent to the node and merge
+        // them into a delta group
+        var deltas = Map.empty[String, ReplicatedData]
+        var versions = Map.empty[String, (Long, Long)] // only for debug so far
```

Comment:
Maybe remove it for now, since it is only collected but never reported/used anywhere ?